Initial commit for core-edge

mneedham opened the pull request:
## catchup

 Store copy and transaction pulling, similar to the way slaves catch up with
 master in the traditional HA setup.
## discovery

Uses Hazelcast to discover an initial cluster topology both for Core and Edge
instances. This discovery layer is used to prompt RAFT cluster changes in Core as
instances leave and join, and used as a topology manager for Edge instances.
## raft

An implementation of a distributed log kept consistent by RAFT. While there
are several packages in here, the key ones are:
- log - where the log implementations for the RAFT protocol are kept
- replication - machinery for replicating entries in the raft log
- membership - where RAFT group membership is maintained (see above for
  initial setup)
## server

 The server package is where all the code is brought together. It also provides
 some shared infrastructure for sending messages, dealing with settings etc.

lutovich commented:
Holy shit! This is impossible to review! Can it be at least split into multiple commits?

davidegrohmann commented:
It would have been nice to have preserved history for future reference. What is the reason for having squashed everything into a single commit?

davidegrohmann commented on community/kernel/pom.xml:
```diff
@@ -134,7 +134,7 @@ the relevant Commercial Agreement.
           <dotGitDirectory>${git.directory}</dotGitDirectory>
           <failOnNoGitDirectory>false</failOnNoGitDirectory>
-          <useNativeGit>true</useNativeGit>
+          <useNativeGit>false</useNativeGit>
```
This is actually a very good example of a change that should have been in a separate commit since there is no way for me to understand why this change has been made.

chrisvest replied:
I think it caused problems for @jimwebber, for some reason, so @digitalstain said he'd make this change and see if it caused any problems. Apparently it hasn't for anyone in the core-edge team, since it's still there.

davidegrohmann replied:
I am a bit lost here, let me understand. Are you saying that this is actually a mistake?

martinfurmanski replied:
This will use the embedded jgit instead of whatever git happens to be available in your environment.

"As rule of thumb - stay on `jgit` (keep this `false`) until you notice performance problems."

davidegrohmann replied:
Right, it sounds reasonable. Can we please have this change in its own PR?
This is completely unrelated to core-edge.

mneedham replied:
@davidegrohmann here we go: https://github.com/neo4j/neo4j/pull/5948

davidegrohmann replied:
@mneedham thanks, shall we remove then that change from this PR?

mneedham replied:
@davidegrohmann good idea. Done.

srbaker commented:
A single, 10,000 line commit? How did you work for so long without committing and not losing anything?

Commit:
Initial commit for core-edge

catchup
-------
 Store copy and transaction pulling, similar to the way slaves catch up with
 master in the traditional HA setup.

discovery
---------
Uses Hazelcast to discover an initial cluster topology both for Core and Edge
instances. This discovery layer is used to prompt RAFT cluster changes in Core as
instances leave and join, and used as a topology manager for Edge instances.

raft
----
An implementation of a distributed log kept consistent by RAFT. While there
are several packages in here, the key ones are:
  * log - where the log implementations for the RAFT protocol are kept
  * replication - machinery for replicating entries in the raft log
  * membership - where RAFT group membership is maintained (see above for
  initial setup)

server
------
 The server package is where all the code is brought together. It also provides
 some shared infrastructure for sending messages, dealing with settings etc.

Commit:
Updates to Neo4j server to introduce core-edge

* Adds CORE and EDGE modes
* Adds endpoints to the HTTP API to distinguish
  leader/follower
